### ISSUE TITLE
Sync BatchNorm

### ISSUE THREAD

**Issue by waychin-weiqin** [feihuzhang/GANet#23]
Hi, I am trying to write a code for semantic segmentation that requires sync BatchNorm and I am interested in the one you used in your code. Can you point me to where can I find more information about the BatchNorm that you have used? 

I tried installing torch-encoding and followed the instructions:

```
 Examples:
        >>> m = SyncBatchNorm(100)
        >>> net = torch.nn.DataParallel(m)
        >>> output = net(input)
```

but my code was stuck at the very beginning. Have you encountered any similar issue? 

Thank you

**Comment by feihuzhang** [comment]
You mean the training freezes after only a few iterations?
I remember sometimes it happens on my centos machine. But, I never saw this on ubuntu + CUDA-10.0 + 2080 GPU. I think it's possibly caused by CUDA or NVIDIA-driver? I am not sure.

Another bug for this sync_bn is that each small batch must be full in each iteration. So you must use drop_last=True when loading the dataset. 

The new offical pytorch is developing sync_bn. Maybe you can try:
https://pytorch.org/docs/master/nn.html#torch.nn.SyncBatchNorm

**Comment by waychin-weiqin** [comment]
Thank you for your reply. 

I was trying to check if the SyncBatchNorm function work so I created a tensor using `torch.randn(<size>)`  and pass through the SyncBatchNorm function. Example: 
`import torch
input = torch.randn(2,32,128,128)
m = SyncBatchNorm2d(32)
output = m(input)`

It got stuck at `output = m(input)`. 

Also, can you elaborate what do you mean by each small batch must be full in each iteration?

Thank you so much.

**Comment by feihuzhang** [comment]
Ok, we are talking on different things.
For sementic segmentation, I think you can refer to this:
https://github.com/hszhao/semseg

**Comment by waychin-weiqin** [comment]
This is what I needed! Thank you very much!
